src/hash-stream-validator.ts

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {createHash, Hash} from 'crypto';
+import {Transform} from 'stream';
+
+import {
+  CRC32CValidatorGenerator,
+  CRC32C_DEFAULT_VALIDATOR_GENERATOR,
+  CRC32CValidator,
+} from './crc32c';
+
+interface HashStreamValidatorOptions {
+  crc32c: boolean;
+  md5: boolean;
+  crc32cGenerator: CRC32CValidatorGenerator;
+}
+
+class HashStreamValidator extends Transform {
+  readonly crc32cEnabled: boolean;
+  readonly md5Enabled: boolean;
+
+  #crc32cHash?: CRC32CValidator = undefined;
+  #md5Hash?: Hash = undefined;
+
+  #md5Digest = '';
+
+  constructor(options: Partial<HashStreamValidatorOptions> = {}) {
+    super();
+
+    this.crc32cEnabled = !!options.crc32c;
+    this.md5Enabled = !!options.md5;
+
+    if (this.crc32cEnabled) {
+      const crc32cGenerator =
+        options.crc32cGenerator || CRC32C_DEFAULT_VALIDATOR_GENERATOR;
+
+      this.#crc32cHash = crc32cGenerator();
+    }
+
+    if (this.md5Enabled) {
+      this.#md5Hash = createHash('md5');
+    }
+  }
+
+  _flush(callback: () => void) {
```

Comment:
Might be good to also just mark these functions with `private`.

Reply:
Wanted to, but TS doesn’t like it as the parent class has it `public`